osidb/models.py

```diff
@@ -1952,6 +1952,83 @@ def bzsync(self, *args, bz_api_key=None, **kwargs):
         self.flaw.save(*args, bz_api_key=bz_api_key, **kwargs)
 
 
+class FlawAcknowledgmentManager(ACLMixinManager, TrackingMixinManager):
+    """flaw acknowledgment manager"""
+
+    @staticmethod
+    def create_flawacknowledgment(
+        flaw, name, affiliation, from_upstream, **extra_fields
+    ):
+        """return a new flaw acknowledgment or update an existing flaw acknowledgment without saving"""
+        try:
+            flawacknowledgment = FlawAcknowledgment.objects.get(
+                flaw=flaw,
+                name=name,
+                affiliation=affiliation,
+                from_upstream=from_upstream,
+            )
+            for attr, value in extra_fields.items():
+                setattr(flawacknowledgment, attr, value)
+            return flawacknowledgment
+
+        except ObjectDoesNotExist:
+            return FlawAcknowledgment(
+                flaw=flaw,
+                name=name,
+                affiliation=affiliation,
+                from_upstream=from_upstream,
+                **extra_fields,
+            )
+
+
+class FlawAcknowledgment(ACLMixin, BugzillaSyncMixin, TrackingMixin):
+    """Model representing flaw acknowledgments"""
+
+    # internal primary key
+    uuid = models.UUIDField(primary_key=True, default=uuid.uuid4, editable=False)
+
+    name = models.CharField(
+        max_length=255
+    )  # max length seen in production as of 02/2023 == 122
+    affiliation = models.CharField(
+        max_length=255
+    )  # max length seen in production == 86
+    from_upstream = models.BooleanField()
```

Comment:
can we document what these two are supposed to mean?

Reply:
Not sure if @jsvob is aware :smile: I myself have only a vague understanding. Affiliation is a company or an organization which the reporter is reporting the vulnerability on behalf (probably makes me think that it should be optional as some people may report on their own). I am not sure what exactly means from upstream. I guess that the person is part of RH consumed upstream project. I would ask either @mprpic or someone from IR team like Marian.

Reply:
I thought there's no hidden complexity but I don't know any more details :D The [relevant part of the schema](https://github.com/RedHatProductSecurity/osidb/blob/fc93883d9e7d1a531b15eeb4e1f019dd2bd7f843/apps/bbsync/srtnotes-schema.json#L113) seemed very straightforward:

```
    "acknowledgment": {
      "type": "object",
      "description": "A single acknowledgment",
      "required": ["name", "affiliation", "from_upstream"],
      "properties": {
        "name": {
          "type": "string",
          "description": "The name of the person or entity being acknowledged."
        },
        "affiliation": {
          "type": ["string", "null"],
          "description": "The affiliation of the person being acknowledged."
        },
        "from_upstream": {
          "type": "boolean",
          "description": "Whether this acknowledgment comes an upstream source."
        }
      }
```

As for _upstream_, that's a term we've been using frequently and I also don't see any hidden complexity:

> [(...) upstream in open source is the source repository and project where contributions happen and releases are made (...)](https://www.redhat.com/en/blog/what-open-source-upstream)

So it all seems that the `from_upstream` bool labels `affiliation`. And I guess that if an upstream of project A discovers a flaw in otherwise-unrelated project B, then the given `affiliation` string would be labeled with a `False` `from_upstream` in that acknowledgement, whereas in an unrelated flaw in project A discovered by someone from the same project, there might be another acknowledgement with the same `affiliation` string but a `True` `from_upstream`. Right?

Schema explanation added as comments in an amended commit.

Reply:
`from_upstream` means that the original acknowledgment came from an upstream project. It allows us to create statements such as: "Red Hat would like to thank Bob for reporting this issue to us; the upstream project XYZ acknowledges Alice as the original reporter". This would translate to two acks:

- Name: Bob
- Name: Alice; from_upstream: yes

I think it would be best to align to an existing standard for acknowledgment data, either:

https://github.com/CVEProject/cve-schema/blob/20a9e977d9020c12d7dce07eb7ef8de30bd61f64/schema/v5.0/CVE_JSON_5.0_schema.json#L971
OR
https://docs.oasis-open.org/csaf/csaf/v2.0/os/csaf-v2.0-os.html#3115-acknowledgments-type---example

We already track names and organizations, and the rest of the data can either be expressed as a defined list of types (as in the CVE v5 schema) or a free-form text summary with an optional URL. That would cover many more relationship types than just the "upstreamness" of an acknowledgment.

Reply:
Thank you, @mprpic !

- It seems neither of the existing standards allow for data that would produce "the upstream project XYZ acknowledges Alice as the original reporter". Right?
- In that case, whatever we go with, we'll still need the `from_upstream` bool, or to have an additional free-form field that captures who expresses the acknowledgment, right?
- Which standard shall we choose? @osoukup @Elkasitu
- Should that happen now? It seems to me it would have to be stored elsewhere than in the existing srtnotes["acknowledgments"].

Reply:
> * It seems neither of the existing standards allow for data that would produce "the upstream project XYZ acknowledges Alice as the original reporter". Right?
> 

Why not? If we chose CVEv5 types, the original reporter (upstream one) would be tagged with `finder` and whoever notified us as `reporter`. If we used the summary field from the CSAF spec, it could be represented there in textual form and left to the analyst (or if we want to handhold a bit more, using static values).

Reply:
>     * Which standard shall we choose? @osoukup @Elkasitu

Nothing new was requested. We have no requirements from the IR team other than do what SFM2 does so I would stick with it and just implement the counterpart in OSIDB. What is described above sounds interesting and might be the future but I would consider it out of scope of this task and also out of scope of SFM2 decommission work.
 
>     * Should that happen now? It seems to me it would have to be stored elsewhere than in the existing srtnotes["acknowledgments"].

Let us stay backwards compatible for now so SFM2 and OSIDB understand each other.